### DISCUSSION TITLE
Best practices, linting standards, & style guide

### DISCUSSION THREAD

**Discussion by ChrisShank** [statelyai/xstate#2354] [General]
If you were to build out a set of linting standards or a style guide for xstate that help enforce best practices what would you include? This could be anything from how you order properties on a `StateNode` to making sure services, actions, and guards are always externalized.

**Comment by robertpenner** [comment]
## Ordering of machine config

Because XState is all about top-down, I place a state node's "own stuff" before its child states. ID first (if it exists, but try to make it unnecessary). Then context if on the root node. 

Then the node's own events: first `onDone` (if there is one) because it's special and great for encapsulation. 

Then the regular events in `on`. A `BACK` event is quite common and high-level so I put that first. 

(As for the state's `entry` and `exit` actions, I'm on the fence as to where they should go in the order.)

Then the child stuff: `initial`, then the child `states`, with the initial state written first, of course, and final state written last (unless there are multiple final states and it reads better to have each branch to a final state as its own sequence). In other words, don't have stuff in-between `initial` and the state node it refers to. 

In the "extra stuff" object, I put action implementations first because they're the most likely to be needed. Then services because they're similar to actions. Then guards because I tend to add them later. (I don't have strong preferences or rationales here.)

My meta-rule is to try to tell a coherent story that flows. Sometimes that means breaking global consistency to optimize locally.

**Reply by ChrisShank** [reply]
I think there are a couple of levels of ordering to look at. I haven't put much thought into this, but this is my rule of thing right now:

1. `StateNode`
  I tend to order properties on a state node to match the lifecycle they are interpreted as so this is typically the order:

```ts
createMachine({
  id: 'someMachine',
  context: {...},
  entry: 'someAction',
  invoke: 'someService', // root level invoke
  initial: 'someState', // keep initial next to states
  states: {
    someParallelState: {
      type: 'parallel', // keep type close to the top, but also close to the properties that it determines
      states: {...},
    },
    someInvokedState: {
      entry: 'someAction',
      invoke: {
        src: 'someService',
        onDone: 'someDoneTarget',
        onError: 'someErrorTarget',
      },
      on: { // events are after invoke especially when invoking a callback that is communicating back to this machine
        someEvent: 'someTarget'
      },
      exit: 'someAction',
    }
  },
  // events happen after hierarchical states so we should 
  on: {
    someEvent: {cond: 'someCondition', target: 'someTarget', actions: 'someAction'} // keep conditions in front of target
  },
  exit: 'someAction,
})
```

2.  `StateNodes`
I typically order state nodes to *closely* match their order of execution. For example the initial state should always be first and final states are typically last. This will render a little better in the current visualizer and helps keep the code match the visualization best.

**Reply by robertpenner** [reply]
Nice work filling in more of the state node lifecycle.

In terms of your statement that "events happen after hierarchical states," maybe that's literally true but I would say temporality is not the only axis that matters. The hierarchy axis is significant; I would argue it's more significant. 

The code layout is also arguably cleaner to have get the parent's complete lifecycle and events declared before moving to the child nodes. Whereas if the `on` and `exit` go after the child `states`, then with multiple levels of nesting, more and more child stuff gets in-between the `on` and `exit` and the rest of the state node they belong to.

**Comment by robertpenner** [comment]
## Statechart Identifiers 

### Natural language with spaces

I have enjoyed abandoning camelCase and SCREAMING_SNAKE_CASE for any identifiers that show up in the statechart. It really humanizes it to have relaxed language with spaces. At first I thought using quotes on object keys everywhere would be annoying, but it wasn't. It was well worth it to make the statechart more approachable for colleagues in Product, Design, QA, etc. 

### Events natural but distinct

For events though, I still use ALL CAPS, but with spaces to feel natural. The caps are a very practical convention to immediately know that something is an event. (Plus the all caps reads nicely in the tiny buttons generated for the events.)

### Events as verbs

For event grammar, I use present imperative tense as much as possible; it matches DOM events in saying CLICK instead of CLICKED. Also, CLICKED is adjectivey, so it sounds more like a state than CLICK does.

### Toggle states as non-Booleans

For toggle states, so far I have avoided using "true" and "false" strings as values. I feel like they're not wrong per se, but they could be confused with Boolean values, and then the string "false" being truthy could lead to bugs (you're counting on other developers always matching against `"false"` instead of `false`.

I'm also finding that avoiding "true" and "false" as a discipline leads to more descriptive code. For example, instead of
```
minimized: "true" | "false" 
```
when I use:
```
size: "minimized" | "maximized" 
```
It's more explicit about what the not-minimized state means. And it leaves the door open to additional state values, whether a "medium" size, or transitional animation states like "minimizing" and "maximizing".

**Comment by robertpenner** [comment]
## Encapsulate with onDone 

Embrace `onDone` and final states to make state nodes more modular and reusable by a parent node. 

In this style, the child node knows less about the outside world. It doesn't care which particular state comes after it. It simply says, "I reached my final state; I'm done" and releases control back to the parent. 

The parent decides how to connect a child's `onDone` to a sibling state, or whether the parent itself is done (if the child state was its final state) and so it releases control to the parent's parent, and so on. 

This style produces lovely fractal "Inception kick" structures. The pieces can be more easily reused and reconfigured by rewiring `onDone` handlers.

## Parallel processes with onDone

Already covered in docs but wanted to mention:

https://xstate.js.org/docs/guides/final.html#parallel-states

**Comment by ChrisShank** [comment]
## Dependency Injection

How do you inject services, actions and guards into machines? What about the machine's initial context. Do you `machine.withConfig`/`machine.withContext` or using a factory function? The answer could be different for different use cases! For example I almost always use factory functions as there are more explicit that the machine needs something and is more type safe. `withConfig` can be really useful when testing!

**Reply by tyler-morrison** [reply]
If machines rely on some form of DI, I've started suggesting my team create fallback errors.

This should prevent confusion if someone tweaks `withConfig` options without understanding all of the ramifications.
```javascript
export function updateLocalStorage() {
  throw new Error(`ERROR: Missing Configuration
The 'updateLocalStorage' action needs to be configured its consumer.

SEE: https://xstate.js.org/docs/packages/xstate-react/#configuring-machines`);
}

export const todoListMachine = createMachine(
  config,
  {
    actions: {
      updateLocalStorage
    }
  }
);
```

**Comment by ChrisShank** [comment]
## Inline vs externalize config

In what cases do you tend to inline services, actions and guards or do you tend to externalize them into the config?

**Reply by tyler-morrison** [reply]
90% of the time, I recommend externalizing configs. Inlining is useful for my "rough drafts", but externalizing has a greater ROI for projects with large teams / long lifespans.

**KEY BENEFITS:**
* Legibility – Serialized guards, services, actions are easier to read and visualize.
* Scalability – You can easily refactor complex logic into separate files.

Compare the example [todosMachine.js](https://codesandbox.io/s/xstate-todomvc-33wr94qv1?file=/todosMachine.js) with the serialized version below…
 
```javascript
// NOTE: This is an extreme example where 
import {
  clearCompleted,
  deleteTodo,
  filterList,
  markAllActive,
  markAllCompleted,
  rehydrateList,
  spawnNewTodo,
  updateLocalStorage,
  updateTodo,
  updateTodoInput,
} from './actions';
import { shouldCommitTodo } from './guards';

const todoListMachine = createMachine(
  {
    id: 'todos',
    context: todoListModel.initialContext,
    initial: 'loading',
    states: {
      loading: {
        entry: rehydrateList,
        always: 'ready',
      },
      ready: {},
    },
    on: {
      'NEW_TODO.CHANGE': {
        actions: 'updateTodoInput',
      },
      'NEW_TODO.COMMIT': {
        cond: 'shouldCommitTodo',
        actions: ['spawnNewTodo', 'updateLocalStorage'],
      },
      'TODO.COMMIT': {
        actions: ['updateTodo', 'updateLocalStorage'],
      },
      'TODO.DELETE': {
        actions: ['deleteTodo', 'updateLocalStorage'],
      },
      SHOW: {
        actions: 'filterList',
      },
      'MARK.completed': {
        actions: 'markAllCompleted',
      },
      'MARK.active': {
        actions: 'markAllActive',
      },
      CLEAR_COMPLETED: {
        actions: ['clearCompleted', 'updateLocalStorage'],
      },
    },
  },
  {
    actions: {
      clearCompleted,
      deleteTodo,
      filterList,
      markAllActive,
      markAllCompleted,
      rehydrateList,
      spawnNewTodo,
      updateTodo,
      updateTodoInput,
      updateLocalStorage,
    },
    guards: {
      shouldCommitTodo,
    },
  }
);
```

**Comment by fraserdrops** [comment]
Personally, I'm wary of style guides, and of lint rules that are style warnings rather than compile errors. I used to be in the camp of enforcing style guides, but I would probably only use one now if the fixes could be automated (like Prettier). I'm talking here about naming conventions and the like, choices that don't really impact the implementation.


I see the most value in developing the best practices and patterns around modelling the statecharts themselves, things like when to use parallel states, using conditionals vs a separate transient state, avoiding child states targeting parent level states etc. These are the areas that really bog me down when I'm coding up a statechart, and IMO would benefit the most from having examples and best practices.

**Reply by robertpenner** [reply]
I agree that higher-level best practices are more valuable overall, and nitpicky lint rules can be annoying or distracting. However, that doesn't negate the value of style guides to make code more consistent and legible. 

To make a reductio regarding "choices that don't really impact the implementation": if we run JavaScript through a minifier, the implementation isn't impacted, but the minified code isn't ergonomic for humans.

**Comment by robertpenner** [comment]
Use trailing commas wherever possible (it's an easy Prettier config). Trailing commas make machine config object easier to rearrange (and reduces diff size when appending code).
